core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -55,6 +63,8 @@
   private boolean sendHeadersCalled;
   private boolean closeCalled;
   private Compressor compressor;
+  private boolean messageSent;
+  private boolean internalClosed;
```

Comment:
Did you discover `internalClosed` was necessary? I think it should be unnecessary.